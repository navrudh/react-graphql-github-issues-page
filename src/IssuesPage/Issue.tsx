import * as React from 'react';
import styled from 'styled-components';
import { IssueNode } from './IssuesList';

const Title = styled.h1.attrs({
  className: 'georgia f4 fw4 mt0 mb0 black-60'
})``;

const Wrapper = styled.article.attrs({className: 'mw7 center bg-washed-blue br3 pa3 pa4-ns mv3 ba b--black-10'})``;

class Issue extends React.Component<IssueNode> {
  private issue: IssueNode;

  constructor(prop: IssueNode) {
    super(prop);
    this.issue = prop;
  }

  public render() {
    return (
      <Wrapper>
        <Title> {this.issue.title} </Title>
        {this.issue.labels.edges.map(({ node }) => (
          <div key={node.name}>
            <p>{`${node.name}: ${node.color}`}</p>
          </div>
        ))}
      </Wrapper>
    );
  }
}

export default Issue;
