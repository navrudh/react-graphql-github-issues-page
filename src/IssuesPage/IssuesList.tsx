import * as React from 'react';

import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Issue from './Issue';

interface QueryResult {
  repository: Repository;
}

interface Repository {
  issues: Issues;
}

interface Issues {
  totalCount: number;
  edges: IssueEdge[];
}

interface IssueEdge {
  node: IssueNode;
}

export interface IssueNode {
  id: string;
  title: string;
  url: string;
  labels: IssueLabel;
}

interface IssueLabel {
  edges: LabelEdges[];
}

interface LabelEdges {
  node: LabelNode;
}

interface LabelNode {
  name: string;
  color: string;
}

const query = (repoOwner: string, repoName: string, issuesFirst: number, issueStatus: 'OPEN' | 'CLOSED') => gql`
{
  repository(owner: ${repoOwner}, name: ${repoName}) {
    issues(first: ${issuesFirst}, states: ${issueStatus}) {
      totalCount
      edges {
        node {
          id
          title
          url
          labels(first: 5) {
            edges {
              node {
                name
                color
              }
            }
          }
        }
      }
    }
  }
}

`;

interface IssuesQueryProps {
  repoOwner: string;
  repoName: string;
  issuesFirst?: number;
  issueStatus?: 'OPEN' | 'CLOSED';
}

class IssuesList extends React.Component<IssuesQueryProps> {
  private repoOwner: string;
  private repoName: string;
  private issuesFirst: number;
  private issueStatus: 'OPEN' | 'CLOSED';

  constructor(props: IssuesQueryProps) {
    super(props);
    props = props;
    this.repoOwner = props.repoOwner;
    this.repoName = props.repoName;
    this.issuesFirst = props.issuesFirst || 20;
    this.issueStatus = props.issueStatus || 'OPEN';
  }

  public render() {
    return (
      <Query query={query(this.repoOwner, this.repoName, this.issuesFirst, this.issueStatus)}>
        {({ loading, error, data }) => {
          if (loading) {
            return <p>Loading...</p>;
          }
          if (error) {
            return <p>Error :(</p>;
          }

          const queryResult: QueryResult = data;
          return queryResult.repository.issues.edges.map(({ node }) => <Issue {...node} />);
        }}
      </Query>
    );
  }
}

export default IssuesList;
