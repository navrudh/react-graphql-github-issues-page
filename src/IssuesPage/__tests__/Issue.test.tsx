import * as React from 'react';

import Issue from '../Issue';
import { IssueNode } from '../IssuesList';
import { shallow } from 'enzyme';

const node: IssueNode = {
  id: '1',
  labels: { edges: Array(5).fill({ node: { name: 'fix', color: '000000' } }) },
  title: 'Issue title',
  url: 'link'
};

describe('Issue', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Issue {...node} />);
    expect(wrapper).toMatchSnapshot();
  });
});
