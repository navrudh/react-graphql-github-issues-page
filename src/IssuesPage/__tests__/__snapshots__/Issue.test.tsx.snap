// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue renders correctly 1`] = `
<styled.article>
  <styled.h1>
     
    Issue title
     
  </styled.h1>
  <styled.article
    color="#000000"
    key="fix"
  >
    fix: 000000
  </styled.article>
  <styled.article
    color="#000000"
    key="fix"
  >
    fix: 000000
  </styled.article>
  <styled.article
    color="#000000"
    key="fix"
  >
    fix: 000000
  </styled.article>
  <styled.article
    color="#000000"
    key="fix"
  >
    fix: 000000
  </styled.article>
  <styled.article
    color="#000000"
    key="fix"
  >
    fix: 000000
  </styled.article>
</styled.article>
`;
